tt_eager/tt_dnn/op_library/update_cache/kernels/compute/update_cache.cpp

```diff
@@ -21,45 +20,52 @@ void MAIN {
     constexpr uint32_t num_batched_heads = get_compile_time_arg_val(6);
     constexpr uint32_t Wt = get_compile_time_arg_val(7);
     constexpr uint32_t u_range = get_compile_time_arg_val(8);
+    constexpr uint32_t granularity = get_compile_time_arg_val(9);
+
 
-    untilize_init(in_cb, untilized_in_cb);
+    pack_untilize_init<Wt>(in_cb, untilized_in_cb);
 
     for (uint32_t  h = 0; h < num_batched_heads; ++h) {
-        untilize_init_short(in_cb);
+
         cb_wait_front(in_cb, Wt);
         cb_reserve_back(untilized_in_cb, Wt);
-        untilize_block(in_cb, Wt, untilized_in_cb);
+        pack_untilize_block<Wt>(in_cb, 1, untilized_in_cb);
         cb_push_back(untilized_in_cb, Wt);
         cb_pop_front(in_cb, Wt);
-        untilize_uninit(in_cb);
 
         unpack_reconfig_data_format_srca(in_cb, cache_cb);
+        for(uint32_t u = 0; u < u_range / granularity; ++u) {
+            pack_untilize_init_short<Wt>(cache_cb, untilized_cache_cb);
```

Comment:
Pass in `u_range / granularity` as compile-time arg (division is expensive in RISC-V). And I think we don't need `u_range` anymore?

Also, is `u_range` just the Ht dim? If so, you can call `u_range / granularity` and `granularity` something like: `num_blocks_h` and `block_h`

Reply:
u_range is the real size of the batch, which is dim 0 of the cache in DRAM. 
Jack added it because sometimes our cache is only 8 or 16 users.

How about I make `u_range / granularity` a constexpr? That way it's not confusing what its doing, and it's not a perf hit.

Reply:
Oh u_range is a runtime arg so I can't constexpr.

Reply:
I see no difference in perf when I remove this division. 
I think it's optimized since granularity is always 2 or 1, so compiler must use shifting.

Reply:
`constexpr uint32_t u_range = get_compile_time_arg_val(8);`

Isn't this compile time?

Reply:
only for the compute kernel. Weirdly, reader/writer use runtime args for u_range

Reply:
As long as the arg doesn't vary per core or needs to be updated in callback, it can be compile time. It's pretty arbitrary what is/isn't compile in these kernels.